packages/sitecore-jss/src/personalization/layout-personalization-service.ts

```diff
@@ -0,0 +1,220 @@
+import { ComponentRendering, PersonalizableComponentRendering, RouteData } from '../layout/models';
+import {
+  PersonalizationDecisionData,
+  PersonalizationDecisionsService,
+} from './personalization-decisions-service';
+import { LayoutFragmentService } from './layout-fragment-service';
+import { LayoutPersonalizationUtils } from './layout-personalization-utils';
+
+export interface PersonalizationResult {
+  /**
+   * A value that indicates whether a route is personalizable
+   */
+  isPersonalizable: boolean;
+}
+
+export interface PersonalizationContext {
+  /**
+   * The route path
+   */
+  routePath: string;
+  /**
+   * The language
+   */
+  language: string;
+}
+
+/**
+ * The layout personalization service that provides the asynchronous API
+ * to initiate the personalization and load personalized components
+ */
+export class LayoutPersonalizationService {
+  private personalizationResult: {
+    personalizationOperation?: Promise<{
+      [key: string]: ComponentRendering | null;
+    }>;
+    components?: { [key: string]: ComponentRendering | null };
+  } = {};
+  private layoutPersonalizationUtils = new LayoutPersonalizationUtils();
+
+  constructor(
+    private personalizationDecisionsService: PersonalizationDecisionsService,
+    private layoutFragmentService: LayoutFragmentService
+  ) {}
+
+  /**
+   * Fetches the personalized data.
+   * @param {PersonalizationContext} context The context.
+   * @param {RouteData} route The route.
+   * @returns {PersonalizationResult} The personalization result.
+   */
+  async fetchPersonalization(
+    context: PersonalizationContext,
+    route: RouteData
+  ): Promise<PersonalizationResult> {
+    // clear personalization before getting new one
+    this.personalizationResult = {};
+
+    const personalizedRenderings = this.layoutPersonalizationUtils.getPersonalizableComponents(
+      route.placeholders
+    );
+
+    if (!personalizedRenderings.length) {
+      return { isPersonalizable: false };
+    }
+
+    const currentResult = this.personalizationResult;
+    currentResult.personalizationOperation = this.personalizeComponents(
+      {
+        routePath: context.routePath,
+        language: context.language,
+      },
+      personalizedRenderings
+    );
+
+    try {
+      const components = await currentResult.personalizationOperation;
+      currentResult.components = components;
+      return { isPersonalizable: true };
+    } catch (error) {
+      currentResult.personalizationOperation = undefined;
+      throw error;
+    }
+  }
+
+  /**
+   * Gets the personalized component.
+   * @param {string} componentUid The unique identifier of a component.
+   * @returns {ComponentRendering} The personalized component.
+   */
+  getPersonalizedComponent(componentUid: string): ComponentRendering | null {
+    if (!this.personalizationResult.components) {
+      return null;
+    }
+
+    return this.personalizationResult.components[componentUid] ?? null;
```

Comment:
if the left side is undefined then return null? can we just return the undefined?